mlflow/utils/environment.py

```diff
@@ -1,31 +1,34 @@
+import yaml
+
 from mlflow.utils import PYTHON_VERSION
 
-_conda_header = """name: mlflow-env
+_conda_header = """\
+name: mlflow-env
 channels:
   - anaconda
   - defaults
-dependencies:"""
+"""
 
 
-def _mlflow_conda_env(path, additional_conda_deps=None, additional_pip_deps=None):
+def _mlflow_conda_env(path, additional_conda_deps=None, additional_pip_deps=None,
+                      additional_conda_channels=None):
     """
     Create conda environment file. Contains default dependency on current python version.
     :param path: local filesystem path where the conda env file is to be created.
     :param additional_conda_deps: List of additional conda dependencies passed as strings.
     :param additional_pip_deps: List of additional pip dependencies passed as strings.
-    :return: path where the files has been created
+    :param additional_channels: List of additional conda channels to search when resolving packages.
+    :return: path where the conda environment file has been created.
     """
-    conda_deps = ["python={}".format(PYTHON_VERSION)]
-    if additional_conda_deps:
-        conda_deps += additional_conda_deps
-    pip_deps = additional_pip_deps
+    env = yaml.load(_conda_header)
+    env["dependencies"] = ["python={}".format(PYTHON_VERSION)]
+    if additional_conda_deps is not None:
+        env["dependencies"] += additional_conda_deps
+    if additional_pip_deps is not None:
+        env["dependencies"].append({"pip": additional_pip_deps})
+    if additional_conda_channels is not None:
+        env["channels"] += additional_conda_channels
+
     with open(path, "w") as f:
-        f.write(_conda_header)
-        prefix = "\n  - "
-        f.write(prefix + prefix.join(conda_deps))
-        if pip_deps:
-            f.write(prefix + "pip:")
-            prefix = "\n    - "
-            f.write(prefix + prefix.join(pip_deps))
-        f.write("\n")
+        yaml.dump(env, f, default_flow_style=False)
```

Comment:
Consider using `yaml.safe_dump` here, which will guarantee that we only write standard YAML to the file (i.e. it'll defend against unexpected things being passed to `additional_conda_deps`, etc). From the [PyYaml docs](https://pyyaml.org/wiki/PyYAMLDocumentation):

> safe_dump(data, stream=None) serializes the given Python object into the stream. If stream is None, it returns the produced stream. safe_dump produces only standard YAML tags and cannot represent an arbitrary Python object.

Reply:
Changed to `yaml.safe_dump`. Thanks!